File: SRC/Host/Loader/CodeDomHostLoader.vb
Imports System.ComponentModel.Design
Imports System.ComponentModel.Design.Serialization
Imports System.Collections
Imports System.Diagnostics
Imports System.IO
Imports System.Reflection
Imports System.Windows.Forms
Imports System.CodeDom.Compiler
Imports System.CodeDom
Imports Microsoft.CSharp
Imports Microsoft.VisualBasic

Namespace Loader
    ''' <summary>
    ''' Inherits from CodeDomDesignerLoader. It can generate C# or VB code
    ''' for a HostSurface. This loader does not support parsing a 
    ''' C# or VB file.
    ''' </summary>
    Public Class CodeDomHostLoader
        Inherits CodeDomDesignerLoader

        Private _csCodeProvider As CSharpCodeProvider = New CSharpCodeProvider()
        Private codeCompileUnit As CodeCompileUnit = Nothing
        Private cg As CodeGen = Nothing
        Private _trs As TypeResolutionService = Nothing
        Private executable As String
        Private f_Run As Process

        Public Sub New()
            _trs = New TypeResolutionService()
        End Sub

        Protected Overrides ReadOnly Property TypeResolutionService As ITypeResolutionService
            Get
                Return _trs
            End Get
        End Property

        Protected Overrides ReadOnly Property CodeDomProvider As CodeDomProvider
            Get
                Return _csCodeProvider
            End Get
        End Property


        ''' <summary>
        ''' Bootstrap method - loads a blank Form
        ''' </summary>
        ''' <returns></returns>
        Protected Overrides Function Parse() As CodeCompileUnit
            Dim ccu As CodeCompileUnit = Nothing
            Dim ds As DesignSurface = New DesignSurface()
            ds.BeginLoad(GetType(Form))
            Dim idh = CType(ds.GetService(GetType(IDesignerHost)), IDesignerHost)
            idh.RootComponent.Site.Name = "Form1"
            cg = New CodeGen()
            ccu = cg.GetCodeCompileUnit(idh)
            Dim names As AssemblyName() = Assembly.GetExecutingAssembly().GetReferencedAssemblies()

            For i = 0 To names.Length - 1
                Dim assembly = System.Reflection.Assembly.Load(names(i))
                ccu.ReferencedAssemblies.Add(assembly.Location)
            Next

            codeCompileUnit = ccu
            Return ccu
        End Function


        ''' <summary>
        ''' When the Loader is Flushed this method is called. The base class
        ''' (CodeDomDesignerLoader) creates the CodeCompileUnit. We
        ''' simply cache it and use this when we need to generate code from it.
        ''' To generate the code we use CodeProvider.
        ''' </summary>
        Protected Overrides Sub Write(ByVal unit As CodeCompileUnit)
            codeCompileUnit = unit
        End Sub

        Protected Overrides Sub OnEndLoad(ByVal successful As Boolean, ByVal errors As ICollection)
            MyBase.OnEndLoad(successful, errors)

            If errors IsNot Nothing Then
                Dim ie As IEnumerator = errors.GetEnumerator()

                While ie.MoveNext()
                    Trace.WriteLine(ie.Current.ToString())
                End While
            End If
        End Sub


#Region "Public methods"

        ''' <summary>
        ''' Flushes the host and returns the updated CodeCompileUnit
        ''' </summary>
        ''' <returns></returns>
        Public Function GetCodeCompileUnit() As CodeCompileUnit
            Flush()
            Return codeCompileUnit
        End Function


        ''' <summary>
        ''' This method writes out the contents of our designer in C# and VB.
        ''' It generates code from our codeCompileUnit using CodeRpovider
        ''' </summary>
        Public Function GetCode(ByVal context As String) As String
            Flush()
            Dim o As CodeGeneratorOptions = New CodeGeneratorOptions()
            o.BlankLinesBetweenMembers = True
            o.BracingStyle = "C"
            o.ElseOnClosing = False
            o.IndentString = "    "

            If Equals(context, "C#") Then
                Dim swCS As StringWriter = New StringWriter()
                Dim cs As CSharpCodeProvider = New CSharpCodeProvider()
                cs.GenerateCodeFromCompileUnit(codeCompileUnit, swCS, o)
                Dim code As String = swCS.ToString()
                swCS.Close()
                Return code
            ElseIf Equals(context, "VB") Then
                Dim swVB As StringWriter = New StringWriter()
                Dim vb As VBCodeProvider = New VBCodeProvider()
                vb.GenerateCodeFromCompileUnit(codeCompileUnit, swVB, o)
                Dim code As String = swVB.ToString()
                swVB.Close()
                Return code
            End If

            Return String.Empty
        End Function



#End Region

#Region "Build and Run"

        ''' <summary>
        ''' Called when we want to build an executable. Returns true if we succeeded.
        ''' </summary>
        Public Function Build() As Boolean
            Flush()


            ' If we haven't already chosen a spot to write the executable to,
            ' do so now.
            If Equals(executable, Nothing) Then
                Dim dlg As SaveFileDialog = New SaveFileDialog()
                dlg.DefaultExt = "exe"
                dlg.Filter = "Executables|*.exe"

                If dlg.ShowDialog() = DialogResult.OK Then
                    executable = dlg.FileName
                End If
            End If

            If Not Equals(executable, Nothing) Then
                ' We need to collect the parameters that our compiler will use.
                Dim cp As CompilerParameters = New CompilerParameters()
                Dim assemblyNames As AssemblyName() = Assembly.GetEntryAssembly().GetReferencedAssemblies()

                For Each an In assemblyNames
                    Dim assembly = System.Reflection.Assembly.Load(an)
                    cp.ReferencedAssemblies.Add(assembly.Location)
                Next

                cp.GenerateExecutable = True
                cp.OutputAssembly = executable

                ' Remember our main class is not Form, but Form1 (or whatever the user calls it)!
                cp.MainClass = "DesignerHostSample." & LoaderHost.RootComponent.Site.Name
                Dim cc As CSharpCodeProvider = New CSharpCodeProvider()
                Dim cr = cc.CompileAssemblyFromDom(cp, codeCompileUnit)

                If cr.Errors.HasErrors Then
                    Dim errors = ""

                    For Each [error] As CompilerError In cr.Errors
                        errors += [error].ErrorText & Microsoft.VisualBasic.Constants.vbLf
                    Next

                    MessageBox.Show(errors, "Errors during compile.")
                End If

                Return Not cr.Errors.HasErrors
            End If

            Return False
        End Function


        ''' <summary>
        ''' Here we build the executable and then run it. We make sure to not start
        ''' two of the same process.
        ''' </summary>
        Public Sub Run()
            If f_Run Is Nothing OrElse f_Run.HasExited Then

                If Build() Then
                    f_Run = New Process()
                    f_Run.StartInfo.FileName = executable
                    f_Run.Start()
                End If
            End If
        End Sub


        ''' <summary>
        ''' Just in case the red X in the upper right isn't good enough,
        ''' we can kill our process here.
        ''' </summary>
        Public Sub [Stop]()
            If f_Run IsNot Nothing AndAlso Not f_Run.HasExited Then
                f_Run.Kill()
            End If
        End Sub

#End Region

    End Class ' class
End Namespace ' namespace

File: SRC/OSMLoader/Loader/CodeDomHostLoader.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.ComponentModel.Design.Serialization;
using System.Collections;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Reflection;
using System.Runtime.Serialization.Formatters.Soap;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Windows.Forms;
using System.Xml;
using System.CodeDom.Compiler;
using System.CodeDom;
using System.Drawing;
using Microsoft.CSharp;
using Microsoft.VisualBasic;

namespace OSMLoader.Loader
{
    /// <summary>
    /// Inherits from CodeDomDesignerLoader. It can generate C# or VB code
    /// for a HostSurface. This loader does not support parsing a
    /// C# or VB file.
    /// </summary>
    public class CodeDomHostLoader : CodeDomDesignerLoader
    {
        private CSharpCodeProvider _csCodeProvider = new CSharpCodeProvider();
        private CodeCompileUnit codeCompileUnit = null;
        private CodeGen cg = null;
        private TypeResolutionService _trs = null;
        private string executable;
        private Process f_Run;

        public CodeDomHostLoader()
        {
            _trs = new TypeResolutionService();
        }

        protected override ITypeResolutionService TypeResolutionService
        {
            get
            {
                return _trs;
            }
        }

        protected override CodeDomProvider CodeDomProvider
        {
            get
            {
                return _csCodeProvider;
            }
        }


        /// <summary>
        /// Bootstrap method - loads a blank Form
        /// </summary>
        /// <returns></returns>
        protected override CodeCompileUnit Parse()
        {
            CodeCompileUnit ccu = null;
            var ds = new DesignSurface();
            ds.BeginLoad(typeof(Form));
            IDesignerHost idh = (IDesignerHost)ds.GetService(typeof(IDesignerHost));
            idh.RootComponent.Site.Name = "Form1";
            cg = new CodeGen();
            ccu = cg.GetCodeCompileUnit(idh);
            var names = Assembly.GetExecutingAssembly().GetReferencedAssemblies();
            for (int i = 0, loopTo = names.Length - 1; i <= loopTo; i++)
            {
                var assembly = Assembly.Load(names[i]);
                ccu.ReferencedAssemblies.Add(assembly.Location);
            }

            codeCompileUnit = ccu;
            return ccu;
        }


        /// <summary>
        /// When the Loader is Flushed this method is called. The base class
        /// (CodeDomDesignerLoader) creates the CodeCompileUnit. We
        /// simply cache it and use this when we need to generate code from it.
        /// To generate the code we use CodeProvider.
        /// </summary>
        protected override void Write(CodeCompileUnit unit)
        {
            codeCompileUnit = unit;
        }

        protected override void OnEndLoad(bool successful, ICollection errors)
        {
            base.OnEndLoad(successful, errors);
            if (errors is object)
            {
                var ie = errors.GetEnumerator();
                while (ie.MoveNext())
                    Trace.WriteLine(ie.Current.ToString());
            }
        }


        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        /// <summary>
        /// Flushes the host and returns the updated CodeCompileUnit
        /// </summary>
        /// <returns></returns>
        public CodeCompileUnit GetCodeCompileUnit()
        {
            Flush();
            return codeCompileUnit;
        }


        /// <summary>
        /// This method writes out the contents of our designer in C# and VB.
        /// It generates code from our codeCompileUnit using CodeRpovider
        /// </summary>
        public string GetCode(string context)
        {
            Flush();
            var o = new CodeGeneratorOptions();
            o.BlankLinesBetweenMembers = true;
            o.BracingStyle = "C";
            o.ElseOnClosing = false;
            o.IndentString = "    ";
            if (Equals(context, "C#"))
            {
                var swCS = new StringWriter();
                var cs = new CSharpCodeProvider();
                cs.GenerateCodeFromCompileUnit(codeCompileUnit, swCS, o);
                string code = swCS.ToString();
                swCS.Close();
                return code;
            }
            else if (Equals(context, "VB"))
            {
                var swVB = new StringWriter();
                var vb = new VBCodeProvider();
                vb.GenerateCodeFromCompileUnit(codeCompileUnit, swVB, o);
                string code = swVB.ToString();
                swVB.Close();
                return code;
            }

            return string.Empty;
        }



        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        /// <summary>
        /// Called when we want to build an executable. Returns true if we succeeded.
        /// </summary>
        public bool Build()
        {
            Flush();


            // If we haven't already chosen a spot to write the executable to,
            // do so now.
            if (Equals(executable, null))
            {
                var dlg = new SaveFileDialog();
                dlg.DefaultExt = "exe";
                dlg.Filter = "Executables|*.exe";
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    executable = dlg.FileName;
                }
            }

            if (!Equals(executable, null))
            {
                // We need to collect the parameters that our compiler will use.
                var cp = new CompilerParameters();
                var assemblyNames = Assembly.GetEntryAssembly().GetReferencedAssemblies();
                foreach (var an in assemblyNames)
                {
                    var assembly = Assembly.Load(an);
                    cp.ReferencedAssemblies.Add(assembly.Location);
                }

                cp.GenerateExecutable = true;
                cp.OutputAssembly = executable;

                // Remember our main class is not Form, but Form1 (or whatever the user calls it)!
                cp.MainClass = "DesignerHostSample." + LoaderHost.RootComponent.Site.Name;
                var cc = new CSharpCodeProvider();
                var cr = cc.CompileAssemblyFromDom(cp, codeCompileUnit);
                if (cr.Errors.HasErrors)
                {
                    string errors = "";
                    foreach (CompilerError error in cr.Errors)
                        errors += error.ErrorText + Constants.vbLf;
                    MessageBox.Show(errors, "Errors during compile.");
                }

                return !cr.Errors.HasErrors;
            }

            return false;
        }


        /// <summary>
        /// Here we build the executable and then run it. We make sure to not start
        /// two of the same process.
        /// </summary>
        public void Run()
        {
            if (f_Run is null || f_Run.HasExited)
            {
                if (Build())
                {
                    f_Run = new Process();
                    f_Run.StartInfo.FileName = executable;
                    f_Run.Start();
                }
            }
        }


        /// <summary>
        /// Just in case the red X in the upper right isn't good enough,
        /// we can kill our process here.
        /// </summary>
        public void Stop()
        {
            if (f_Run is object && !f_Run.HasExited)
            {
                f_Run.Kill();
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    } // class
} // namespace
